Allocate zeros

dsheets opened the pull request:
This uses `calloc` under the hood to give the OS a chance to optimize if it tags pages as zeroed.

Please review and let me know how you'd like this expressed/integrated and if you see any issues.

Thanks!

dbuenzli commented on src/ctypes/ctypes.mli:
```diff
@@ -184,15 +184,17 @@ val from_voidp : 'a typ -> unit ptr -> 'a ptr
+(** [allocate t v] allocates a fresh value of type [t], initialises it
+    with [v] and returns its address.  The argument [?finalise], if
+    present, will be called just before the memory is freed.  If
+    [?zero] is true (default false), the memory will be zeroed.  *)
```
I don't get it. why do I care that the memory be zeroed if it is initialized with [v] ?

dsheets replied:
Hmm... indeed that seems worthless and should probably be removed.

dsheets commented:
The CI failure looks like it has something to do with Travis/ARM/qemu/docker.

yallop commented:
Thanks for the careful patch!

I think the additional overhead of `calloc` is probably low enough that we can just use `calloc` all the time, and leave the interface unchanged.

dsheets commented:
Ok, I've updated the PR. Now there is a single squashed commit which notes the change of default in the docstring and internal comments. Also, I've completely removed the use of `caml_stat_alloc` and `caml_stat_free` from the memory stubs. I converted `ctypes_copy_bytes` to use OCaml binding macros because it wasn't and put a check for ENOMEM in its body because the checked-malloc in `caml_stat_alloc` was removed.

yallop commented:
Thanks: that looks good.  There's another call to `Stubs.allocate` that needs updating; it's currently causing the build to fail:

```
File "src/ctypes/ctypes_memory.ml", line 27, characters 65-72:
Error: This expression has type int -> Stubs.managed_buffer
       but an expression was expected of type
         Stubs.managed_buffer = Ctypes_memory_stubs.managed_buffer
Makefile.rules:92: recipe for target '_build/src/ctypes/ctypes_memory.cmo' failed
make: *** [_build/src/ctypes/ctypes_memory.cmo] Error 2
```

dsheets commented:
Oops, sorry about that. I've fixed the uses of `allocate` and an issue with `ctypes_copy_bytes`'s missing `CAMLparam`. Let me know if you're happy with this and I can squash the commits.

yallop commented:
> Let me know if you're happy with this and I can squash the commits.

It looks good to me.  I'll merge once it's squashed.

dsheets commented:
Squished.
